jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/gzip/GzipResponse.java

```diff
@@ -249,136 +271,162 @@ protected void onCompleteFailure(Throwable x)
-            if (_deflaterEntry == null)
+            if (LOG.isDebugEnabled())
+                LOG.debug("GzipBufferCB.process(): _last={}, _buffer={}, _content={}", _last, BufferUtil.toDetailString(_buffer), BufferUtil.toDetailString(_content));
+
+            GZState gzstate = _state.get();
+
+            // Are we finished?
+            if (gzstate == GZState.FINISHED)
             {
                 // then the trailer has been generated and written below.
                 // We have finished compressing the entire content, so
                 // cleanup and succeed.
-                if (_buffer != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_buffer);
-                    _buffer = null;
-                }
-                if (_copy != null)
-                {
-                    getRequest().getComponents().getByteBufferPool().release(_copy);
-                    _copy = null;
-                }
+                cleanup();
                 return Action.SUCCEEDED;
             }
 
             // If we have no buffer
             if (_buffer == null)
             {
-                // allocate a buffer and add the gzip header.
                 _buffer = getRequest().getComponents().getByteBufferPool().acquire(_bufferSize, false);
-                BufferUtil.fill(_buffer, GZIP_HEADER, 0, GZIP_HEADER.length);
+                // Per RFC-1952, GZIP is LITTLE_ENDIAN
+                _buffer.order(ByteOrder.LITTLE_ENDIAN);
+                BufferUtil.flipToFill(_buffer);
+                // Add GZIP Header
+                _buffer.put(GZIP_HEADER, 0, GZIP_HEADER.length);
             }
             else
             {
                 // otherwise clear the buffer as previous writes will always fully consume.
-                BufferUtil.clear(_buffer);
+                BufferUtil.clearToFill(_buffer);
             }
 
-            // If the deflater is not finished, then compress more data.
             Deflater deflater = _deflaterEntry.get();
+
+            return switch (gzstate)
+            {
+                case COMPRESSING -> compressing(deflater, _buffer);
+                case FINISHING -> finishing(deflater, _buffer);
+                default -> throw new IllegalStateException("Unexpected state [" + _state.get() + "]");
+            };
+        }
+
+        private void cleanup()
+        {
+            if (_deflaterEntry != null)
+            {
+                _deflaterEntry.release();
+                _deflaterEntry = null;
+            }
+
+            if (_buffer != null)
+            {
+                getRequest().getComponents().getByteBufferPool().release(_buffer);
+                _buffer = null;
+            }
+        }
+
+        private int getFlushMode()
+        {
+            return _syncFlush ? Deflater.SYNC_FLUSH : Deflater.NO_FLUSH;
+        }
+
+        private Action compressing(Deflater deflater, ByteBuffer outputBuffer)
```

Comment:
some javadoc to say something to the effect of "this method is called by `process` when in COMPRESSING state.  It may call `finishing` to complete the compression.